pkg-r/R/chat_append.R

```diff
@@ -0,0 +1,589 @@
+#' `stream_async` fails), the promise returned by `chat_append` will reject with
+#' the error. If the `chat_append` call is the last expression in a Shiny
+#' observer, Shiny will see that the observer failed, and end the user session.
+#' If you prefer to handle the error gracefully, use [promises::catch()] on the
+#' promise returned by `chat_append`.
+#'
+#' @param id The ID of the chat element
+#' @param response The message or message stream to append to the chat element.
+#'   The actual message content can one of the following:
+#'
+#'   * A string, which is interpreted as markdown and rendered to HTML on
+#'     the client.
+#'     * To prevent interpreting as markdown, mark the string as
+#'       [htmltools::HTML()].
+#'   * A UI element.
+#'     * This includes [htmltools::tagList()], which take UI elements
+#'       (including strings) as children. In this case, strings are still
+#'       interpreted as markdown as long as they're not inside HTML.
+#'
+#' @param role The role of the message (either "assistant" or "user"). Defaults
+#'   to "assistant".
+#' @param icon An optional icon to display next to the message, currently only
+#'   used for assistant messages. The icon can be any HTML element (e.g., an
+#'   [htmltools::img()] tag) or a string of HTML.
+#' @param session The Shiny session object
+#'
+#' @returns Returns a promise that resolves to the contents of the stream, or an
+#'   error. This promise resolves when the message has been successfully sent to
+#'   the client; note that it does not guarantee that the message was actually
+#'   received or rendered by the client. The promise rejects if an error occurs
+#'   while processing the response (see the "Error handling" section).
+#'
+#' @examplesIf interactive()
+#' library(shiny)
+#' library(coro)
+#' library(bslib)
+#' library(shinychat)
+#'
+#' # Dumbest chatbot in the world: ignores user input and chooses
+#' # a random, vague response.
+#' fake_chatbot <- async_generator(function(input) {
+#'   responses <- c(
+#'     "What does that suggest to you?",
+#'     "I see.",
+#'     "I'm not sure I understand you fully.",
+#'     "What do you think?",
+#'     "Can you elaborate on that?",
+#'     "Interesting question! Let's examine thi... **See more**"
+#'   )
+#'
+#'   await(async_sleep(1))
+#'   for (chunk in strsplit(sample(responses, 1), "")[[1]]) {
+#'     yield(chunk)
+#'     await(async_sleep(0.02))
+#'   }
+#' })
+#'
+#' ui <- page_fillable(
+#'   chat_ui("chat", fill = TRUE)
+#' )
+#'
+#' server <- function(input, output, session) {
+#'   observeEvent(input$chat_user_input, {
+#'     response <- fake_chatbot(input$chat_user_input)
+#'     chat_append("chat", response)
+#'   })
+#' }
+#'
+#' shinyApp(ui, server)
+#'
+#' @export
+chat_append <- function(
+  id,
+  response,
+  role = "assistant",
+  icon = NULL,
+  session = getDefaultReactiveDomain()
+) {
+  check_active_session(session)
+  if (!is_string(role)) {
```

Comment:
Previously, we were checking that `role` is either `"assistant"` or `"user"`. Now that we're relying more on role, I've relaxed this constraint. By doing this, we can allow for multiple chat participants in a single chat UI, e.g. each with their own icons and styles, etc.

On the TypeScript side, non-user roles get a robot icon by default. So, in general, we carve out user messages for special styles and treat all others as variants of a system message.